File: ios/App/Query/ListPaginationStore.swift
import Foundation
import SwiftUI

@MainActor
final class ListPaginationStore: ObservableObject {
  @Published var loading = false
  @Published var loadingPage = false
  @Published var pageSize: Int
  @Published var currentPage = 1
  @Published var hasNextPage = false
  @Published var endCursor: String?
  @Published var pageToAfterCursor: [Int: String] = [:]

  private let cursorPagination = CursorPaginationController()

  init(pageSize: Int) {
    self.pageSize = pageSize
    syncFromController()
  }

  func setLoading(resetPagination: Bool, withLoading: Bool) {
    if resetPagination {
      loading = withLoading
      return
    }
    loadingPage = true
  }

  func clearLoading(resetPagination: Bool) {
    if resetPagination {
      loading = false
      return
    }
    loadingPage = false
  }

  func resetPaginationState() {
    cursorPagination.reset()
    syncFromController()
  }

  func afterCursor(forPage page: Int) -> String? {
    cursorPagination.afterCursor(forPage: page)
  }

  func nextPageNumberForLoading() -> Int? {
    guard !loadingPage else {
      return nil
    }
    guard let nextPage = cursorPagination.prepareNextPage() else {
      return nil
    }
    syncFromController()
    return nextPage
  }

  func previousPageNumberForLoading() -> Int? {
    guard !loadingPage else {
      return nil
    }
    return cursorPagination.previousPage()
  }

  func applyPageResult(page: Int, endCursor: String?, serverHasNextPage: Bool) {
    cursorPagination.applyPageResult(page: page, endCursor: endCursor, serverHasNextPage: serverHasNextPage)
    syncFromController()
  }

  private func syncFromController() {
    currentPage = cursorPagination.currentPage
    hasNextPage = cursorPagination.hasNextPage
    endCursor = cursorPagination.endCursor
    pageToAfterCursor = cursorPagination.pageToAfterCursor
  }
}

File: android/app/src/main/kotlin/com/baseapp/android/store/ListPaginationStore.kt
package com.baseapp.android.store

import com.baseapp.android.query.CursorPaginationController
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

data class ListPaginationState(
    val loading: Boolean = false,
    val loadingPage: Boolean = false,
    val currentPage: Int = 1,
    val hasNextPage: Boolean = false,
    val endCursor: String? = null,
)

/**
 * Pagination state container for list feeds (state only; the load workflows
 * live in ListFeedLoader/ListFeedController).
 */
class ListPaginationStore(val pageSize: Int) {
    private val cursorPagination = CursorPaginationController()
    private val _state = MutableStateFlow(ListPaginationState())
    val state: StateFlow<ListPaginationState> = _state.asStateFlow()

    val loading: Boolean get() = _state.value.loading
    val loadingPage: Boolean get() = _state.value.loadingPage
    val currentPage: Int get() = _state.value.currentPage
    val hasNextPage: Boolean get() = _state.value.hasNextPage

    fun setLoading(resetPagination: Boolean, withLoading: Boolean) {
        if (resetPagination) {
            _state.value = _state.value.copy(loading = withLoading)
            return
        }
        _state.value = _state.value.copy(loadingPage = true)
    }

    fun clearLoading(resetPagination: Boolean) {
        if (resetPagination) {
            _state.value = _state.value.copy(loading = false)
            return
        }
        _state.value = _state.value.copy(loadingPage = false)
    }

    fun resetPaginationState() {
        cursorPagination.reset()
        syncFromController()
    }

    fun afterCursor(forPage: Int): String? = cursorPagination.afterCursor(forPage)

    fun nextPageNumberForLoading(): Int? {
        if (loadingPage) {
            return null
        }
        val nextPage = cursorPagination.prepareNextPage() ?: return null
        syncFromController()
        return nextPage
    }

    fun applyPageResult(page: Int, endCursor: String?, serverHasNextPage: Boolean) {
        cursorPagination.applyPageResult(page, endCursor, serverHasNextPage)
        syncFromController()
    }

    private fun syncFromController() {
        _state.value = _state.value.copy(
            currentPage = cursorPagination.currentPage,
            hasNextPage = cursorPagination.hasNextPage,
            endCursor = cursorPagination.endCursor,
        )
    }
}
